ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -1,89 +1,257 @@
+    , testChainRollback       :: WithOrigin SlotNo
+
+      -- | The blocks that 'testChainRollback' would roll back
+    , testChainRollbackBlocks :: [CS.PBftSigner PBftMockCrypto]
+    }
   deriving (Show)
 
+-- | Generate chain state
+--
+-- We don't want to make too many assumptions here, so instead we model what
+-- happens on an actual chain: we have a bunch of slots, starting from genesis,
+-- signed by various keys, some slots are empty. We then compute the maximum
+-- rollback point (@k@ from the tip), and finally drop (d @<= k@) slots,
+-- and take the final @n + k - d@ slots.
+--
+-- Returns both the chain before rollback as well as the chain state.
+--
+-- NOTE: PBftChainState assumes @k >= 1@.
 instance Arbitrary TestChainState where
   arbitrary = do
-    -- The weights and ranges have been tweaked such that
-    -- 'prop_chainStateSize_pruneChainState' can do some pruning.
-    nbKs <- frequency
-      -- We want to test with no keys, but not too often
-      [ (1, return 0)
-      , (9, choose (1, 5))
-      ]
-    nbVs <- frequency
-      -- We want to test with no values, but not too often
-      [ (1, return 0)
-      -- We want to test with fewer values than keys, but not too often
-      , (1, choose (1, nbKs))
-      , (8, choose (nbKs, 200))
-      ]
-    return $ TestChainState . testState Origin $ Map.fromList
-      [ (k, vs)
-      | k <- [0..nbKs-1]
-      , let vs = [v | v <- [0..nbVs-1], v `mod` nbKs == k]
-      ]
-
-testState :: WithOrigin Int -> Map Int [Int] -> PBftChainState PBftMockCrypto
-testState anchor =
-      CS.fromMap (fromIntegral <$> anchor)
-    . Map.mapKeys VerKeyMockDSIGN
-    . Map.map (fmap fromIntegral)
+      k       <- choose (1, 4)
+      n       <- choose (1, 10)
+      numKeys <- choose (1, 4)
+
+      -- Pick number of slots
+      --
+      -- We don't try to be too clever here; we need to test the various edge
+      -- cases near genesis more carefully, but also want to test where we are
+      -- well past genesis
+      numSlots <- oneof [
+          choose (0, k)
+        , choose (0, n)
+        , choose (0, k + n)
+        , choose (0, 5 * (k + n))
+        ]
+
+      -- Generate all the signatures
+      slots <- generateSigners (genMockKey numKeys) numSlots Origin
+
+      -- Compute max rollback point
+      let anchor = case drop (fromIntegral k) (reverse slots) of
+                     []  -> Origin
+                     x:_ -> At (CS.pbftSignerSlotNo x)
+
+      -- Pick a number of blocks to drop
+      toDrop <- choose (0, k)
+      let signers = Seq.fromList $ takeLast (n + k - toDrop) (dropLast toDrop slots)
+          state   = CS.fromList
+                      (SecurityParam k)
+                      (CS.WindowSize n)
+                      (anchor, signers)
+
+      -- Create potential next slot to be added
+      nextKey  <- genMockKey numKeys
+      nextSlot <- case signers of
+                    Seq.Empty ->
+                      SlotNo <$> choose (0, 3)
```

Comment:
Comment: can only empty if near genesis, because `toDrop` is most `k` and hence `signers` must be `n` long, unless `slots` near genesis.